Fix AI panel not closing properly

FadhlanR opened the pull request:
https://github.com/user-attachments/assets/17ebbcc8-6596-4695-aba4-f41eabcbbb3b



In this PR, instead of directly addressing the bug, I decided to refactor resizable-panel by forking logic from the react-resizable-panels library. Now, the bug is fixed and the codes look cleaner.

FadhlanR committed:
Refactor resizablePanelGroup codes by forking some logic from react-resizable-panels

FadhlanR committed:
lint fix

FadhlanR committed:
fix host test

FadhlanR committed:
Update usage

FadhlanR committed:
Update tests

richardhjtan commented:
There is some behaviour change on the resizable panel in code mode. The most noticeable one is that the data section no longer auto-hides when reduced to a certain width.

**Before**

https://github.com/user-attachments/assets/abc72907-8e7e-4998-a3f6-dadcfb934d0d

**Now**

https://github.com/user-attachments/assets/fdecfa27-5eeb-41dc-98a1-d3e65e6cbb5e

It shouldn't be considered a bug but just the decision if we need to preserve this behaviour

richardhjtan commented:
The AI assistant chat max-width has changed, but looks more natural to me

**Before**
![ai-chat-max-width-before](https://github.com/user-attachments/assets/009e8eba-8aba-43ea-985e-e97ee2e9787c)

**After**
![ai-chat-max-width-now](https://github.com/user-attachments/assets/764e466d-56e7-4e7a-b6ee-1be1707a0197)

FadhlanR committed:
Allow code panel to be collapsed

FadhlanR committed:
Merge branch 'main' into cs-7553-fix-ai-panel-not-closing-properly

lukemelia reviewed (changes requested):
File names should be consistent with repository style (dasherized)

lukemelia commented on packages/host/app/components/operator-mode/code-submode.gts:
```diff
@@ -97,21 +96,22 @@ type SelectedAccordionItem = 'schema-editor' | null;
 };
 
 const CodeModePanelHeights = 'code-mode-panel-heights';
 const ApproximateRecentPanelDefaultFraction =
```
If we're going to make this out of 100, rename the constant to `ApproximateRecentPanelDefaultPercentage`

FadhlanR committed:
Rename files

FadhlanR commented:
> There is some behaviour change on the resizable panel in code mode. The most noticeable one is that the data section no longer auto-hides when reduced to a certain width.
> 
> **Before**
> 
>  resize-panel-before.mov 
> **Now**
> 
>  resize-panel-now.mov 
> It shouldn't be considered a bug but just the decision if we need to preserve this behaviour

Thanks for checking, I forgot to remove the collapsible tag for the code panel. I've updated please check it again.

FadhlanR commented:
> The AI assistant chat max-width has changed, but looks more natural to me
> 
> **Before** ![ai-chat-max-width-before](https://private-user-images.githubusercontent.com/30404522/393755977-009e8eba-8aba-43ea-985e-e97ee2e9787c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NjAzNjEsIm5iZiI6MTczMzc2MDA2MSwicGF0aCI6Ii8zMDQwNDUyMi8zOTM3NTU5NzctMDA5ZThlYmEtOGFiYS00M2VhLTk4NWUtZTk3ZWUyZTk3ODdjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA5VDE2MDEwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkNzI5M2IxZDJmZWJhNzAyZTg2MzUxODc0ZDY2ZDNkM2QwMzBiOTQyZjVkZDc4ZWRhMzhkZTdjZDYxNDM4OTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.D0FqaiVqnD7e9j7Yc0ba86QPCvaUTttn08le5meuVVA)
> 
> **After** ![ai-chat-max-width-now](https://private-user-images.githubusercontent.com/30404522/393755987-764e466d-56e7-4e7a-b6ee-1be1707a0197.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NjAzNjEsIm5iZiI6MTczMzc2MDA2MSwicGF0aCI6Ii8zMDQwNDUyMi8zOTM3NTU5ODctNzY0ZTQ2NmQtNTZlNy00ZTdhLWI2ZWUtMWJlMTcwN2EwMTk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA5VDE2MDEwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4M2ZjNmJmMTQxZWFjMTA2NWNmMGFlZTk1NTg0MGQyNzFjZTZjOTZlNGRkYzc1NWRiOWIwMzZjOWE5MjY3YWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2BQ_vjcqVweGpf_AvAWsxIfcv-q2PsZ2p_k62bmxOsc)

I think the previous min size is too small, so I updated it.

FadhlanR commented:
> File names should be consistent with repository style (dasherized)

Updated

burieberry reviewed (commented):
Looks good overall. In terms of ai-chat panel size, I noticed it starts getting cropped from the right when the viewport gets too small and some of the buttons become unreachable. I recommend not letting it get smaller than approximately 300px. The default size of the ai panel in the designs seems to be around 366px.

Buttons cropped on the right-side when too small:
<img width="951" alt="cropped-ai-chat" src="https://github.com/user-attachments/assets/b787dc7a-abf9-44e4-848a-a884691e8e71">

lukemelia reviewed (approved):
The minSize of 25 for the AI assistant panel feels too large on a wide display. Are we limited to percentage-based min sizes?

FadhlanR committed:
Adjust aiPanel widths

FadhlanR commented:
> The minSize of 25 for the AI assistant panel feels too large on a wide display. Are we limited to percentage-based min sizes?

Yes, the pixel-based constraints added significant complexity, which is why we are limited to percentage-based constraints. However, I’ve adjusted the panel widths in submode-layout. It may not work perfectly when resizing the window, but when the AI panel is closed and reopened, it will open with a better size. cc: @burieberry

FadhlanR committed:
Merge branch 'main' into cs-7553-fix-ai-panel-not-closing-properly

lukemelia commented:
I think there used to be a threshold behavior where if you shrank a panel below a certain threshold and it was collapsible it would collapse. Was that intentionally removed?

FadhlanR commented:
> I think there used to be a threshold behavior where if you shrank a panel below a certain threshold and it was collapsible it would collapse. Was that intentionally removed?

It's still there, you can try to resize code editor panel in code submode. It will collapse once the size is lower than minSize.

jurgenwerk commented:
<img width="774" alt="image" src="https://github.com/user-attachments/assets/679d725f-75ca-4d9c-8c7a-cf37039021d4">

Dragging this box is glitchy:

https://github.com/user-attachments/assets/cf8dc269-fc27-4c3a-9777-55f853be8c4d

FadhlanR commented:
> <img width="774" alt="image" src="https://github.com/user-attachments/assets/679d725f-75ca-4d9c-8c7a-cf37039021d4">
> 
> 
> 
> Dragging this box is glitchy:
> 
> 
> 
> https://github.com/user-attachments/assets/cf8dc269-fc27-4c3a-9777-55f853be8c4d
> 
> 
> 
> 
> 
> 

I'll check the box corner.

For the glitch, i think it's expected behavior since we define the minimum size for recent files panel. So the recent files panel will only open if the cursor dragging delta is more than the minimum size. Is it better to remove the minimum size of recent files panel? cc: @lukemelia

jurgenwerk commented:
I think the previous resize behavior of recent files felt much nicer... you could resize it to a small size and when the handle was very close to to the border, it snapped into the closed position.

Another degradation that I found - I would expect to be able to narrow this column down even further but now it snaps closed when there is still plenty of available space for code:

https://github.com/user-attachments/assets/927f927a-25bd-460f-a260-dbb67f1edb98

FadhlanR committed:
Fix box shadow and adjust codeEditor minSize

FadhlanR commented:
> Another degradation that I found - I would expect to be able to narrow this column down even further but now it snaps closed when there is still plenty of available space for code:

I've updated the minimum size of code editor. Please check it again.


> <img alt="image" width="774" src="https://private-user-images.githubusercontent.com/273660/394350660-679d725f-75ca-4d9c-8c7a-cf37039021d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4NDg4NjksIm5iZiI6MTczMzg0ODU2OSwicGF0aCI6Ii8yNzM2NjAvMzk0MzUwNjYwLTY3OWQ3MjVmLTc1Y2EtNGQ5Yy04YzdhLWNmMzcwMzkwMjFkNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxMFQxNjM2MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YzhkM2JhMzVhOTQ3MjAwYjJjNThmODM4ZGVhOTU4ZTU5MTZlZDE0NWMzMzIwMzVhNTY2N2Q0MjQzNWFlYWI4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.qNofNXX9vVGgobYarvM03SUA7JtfdJ6YbJ22Aw9mkps">

Fixed!

jurgenwerk commented:
In my opinion the boxes still snap close too quickly. It feels odd when there is plenty of space left for content and then the box just closes. 

Here is what I am experiencing:

https://github.com/user-attachments/assets/fb5b2b60-0e02-406d-b630-62d1078ea497

Here is the old behaviour where breakpoints aren't that extreme and the experience is nicer: 

https://github.com/user-attachments/assets/d0190242-b09b-4a79-ba8e-43e432c1c95a

FadhlanR commented:
> In my opinion the boxes still snap close too quickly. It feels odd when there is plenty of space left for content and then the box just closes.
> 
> Here is what I am experiencing:
> 
>  Screen.Recording.2024-12-10.at.21.53.29.mov 
> Here is the old behaviour where breakpoints aren't that extreme and the experience is nicer:
> 
>  Screen.Recording.2024-12-10.at.21.59.24.mov

It seems we need to introduce a minimum length in pixels. What do you think if I file separate tickets for the minimum length in pixels and the collapsed panel behavior? I believe the adjustment will take time with the current calculation logic. However, the current logic is much better than the previous one because it fixes ai panel closing issue, provides more precise cursor positioning when dragging the panel, and is more flexible in terms of allowing multiple tables to be resized simultaneously.

I've implemented a workaround to address your feedback regarding collapsed panel behavior [2946b15](https://github.com/cardstack/boxel/pull/1887/commits/2946b15739025ba46546e3389c121d80df4da162)

FadhlanR committed:
Collapsed panel behavior

jurgenwerk commented:
I am not actually seeing any improvement with your workaround (tested in the PR preview link). I am ok if issues are addressed in follow up work.
